### PR TITLE
Information Retrieval (2/5)

### PR DESCRIPTION
**Pull request by lucadiliello**
## Before submitting

- [X] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [X] Did you read the contributor guideline, Pull Request section?
- [x] Did you make sure to update the docs?
- [x] Did you write any new necessary tests?

## What does this PR do?
Implements `Mean Reciprocal Rank` for Information Retrieval. More info [here](https://en.wikipedia.org/wiki/Mean_reciprocal_rank).

## PR review
Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: tests/functional/test_retrieval.py [hunk 1]
```diff
@@ -3,28 +3,83 @@
...
 
 @pytest.mark.parametrize(['sklearn_metric', 'torch_metric'], [
-    pytest.param(sk_average_precision, retrieval_average_precision),
+    [sk_average_precision, retrieval_average_precision],
+    [reciprocal_rank, retrieval_reciprocal_rank],
 ])
-@pytest.mark.parametrize("size", [1, 4, 10, 100])
-def test_against_sklearn(sklearn_metric, torch_metric, size):
-    """Compare PL metrics to sklearn version. """
+@pytest.mark.parametrize("size", [1, 3, 8, 33])
```
Comment (SkafteNicki) [inline tests/functional/test_retrieval.py:19]: any specific reason for these numbers?

File: torchmetrics/retrieval/mean_average_precision.py [hunk 2]
```diff
@@ -61,13 +61,13 @@ class RetrievalMAP(RetrievalMetric):
...
-        >>> target = torch.tensor([False, False, True, False, True, False, False])
+        >>> target = torch.tensor([False, False, True, False, True, False, True])
 
         >>> map = RetrievalMAP()
         >>> map(indexes, preds, target)
-        tensor(0.7500)
+        tensor(0.7917)
         >>> map.compute()
-        tensor(0.7500)
+        tensor(0.7917)
```
Comment (SkafteNicki) [inline torchmetrics/retrieval/mean_average_precision.py:70]: just remove the `map.compute()` part as it is the same

**Review by SkafteNicki — APPROVED** [review]
LGTM

File: tests/functional/test_retrieval.py (hunk 1 above)
Comment (lucadiliello) [reply tests/functional/test_retrieval.py:19]: No, we can safely remove the `33`.

File: torchmetrics/retrieval/mean_average_precision.py (hunk 2 above)
Comment (lucadiliello) [reply torchmetrics/retrieval/mean_average_precision.py:70]: Yes

File: tests/retrieval/test_metrics.py [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+@pytest.mark.parametrize('rounds', [20])
+@pytest.mark.parametrize('n_documents', [1, 5])
+@pytest.mark.parametrize('query_without_relevant_docs_options', ['skip', 'pos', 'neg'])
+def test_against_sklearn(
+    sklearn_metric, torch_metric, size, rounds, n_documents, query_without_relevant_docs_options
+) -> None:
+    """ Compare PL metrics to standard version. """
+    seed_everything(0)
+
+    for _ in range(rounds):
```
Comment (Borda) [inline tests/retrieval/test_metrics.py:27]: why is rounds important, is there a reason that once it can pass but it shall not pass 20x?

File: tests/retrieval/test_metrics.py [hunk 4]
```diff
@@ -0,0 +1,112 @@
+import numpy as np
```
Comment (Borda) [inline tests/retrieval/test_metrics.py:1]: general to the file, lest keep test file per metric unless there is good reason to merge it in one, also "test_metrics" is very bad names assuming that the packages is all about metrics

File: tests/retrieval/test_metrics.py (hunk 3 above)
Comment (lucadiliello) [reply tests/retrieval/test_metrics.py]: No, removed

File: tests/retrieval/test_metrics.py (hunk 4 above)
Comment (lucadiliello) [reply tests/retrieval/test_metrics.py]: Files separated, common logic now in the `helper`.

File: tests/functional/test_retrieval.py [hunk 5]
```diff
@@ -3,28 +3,84 @@
...
+    """ Check wrong input dtypes are managed correctly. """
+    device = 'cuda' if torch.cuda.is_available() else 'cpu'
+    seed_everything(0)
+
+    length = 10  # not important in this case
+    preds = torch.tensor([0] * length, device=device, dtype=torch.float32)
+    target = torch.tensor([0] * length, device=device, dtype=torch.int64)
+
+    # check error on input dtypes are raised correctly
+    with pytest.raises(ValueError):
```
Comment (Borda) [inline tests/functional/test_retrieval.py:58]: pls specify also match error message

**Review by Borda — CHANGES_REQUESTED** [review]

File: tests/functional/test_retrieval.py (hunk 5 above)
Comment (lucadiliello) [reply tests/functional/test_retrieval.py]: Done

File: tests/functional/test_retrieval.py [hunk 6]
```diff
@@ -3,28 +3,95 @@
 import numpy as np
 import pytest
 import torch
+from pytorch_lightning import seed_everything
```
Comment (Borda) [inline tests/functional/test_retrieval.py:6]: don't use any PL features in tests, for seeding use `torch.manuall_seed` and `torch.cuda.manuall_seed_all`

File: docs/source/references/modules.rst [hunk 7]
```diff
@@ -269,4 +262,60 @@ R2Score
...
+    :widths: 10, 10, 10, 10, 10, 10
+
+    "``long``", "(N,...)", "``float``", "(N,...)", "``long`` or ``bool``", "(N,...)"
+
+.. note::
+    All dimensions are "squeezed out" at the beginning, so
+    that, for example, a tensor of shape ``(N, M)`` is treated as ``(N * M, )``.
+
+In Information Retrieval you have a query that is compared with a variable number of documents. For each pair ``(Q_i, D_j)``,
+a score is computed that measures the relevance of document ``D`` wrt query ``Q``. Documents are then sorted by score
```
Comment (SkafteNicki) [inline docs/source/references/modules.rst:288]:
```suggestion
a score is computed that measures the relevance of document ``D`` w.r.t. query ``Q``. Documents are then sorted by score
```

File: docs/source/references/modules.rst [hunk 8]
```diff
@@ -269,4 +262,60 @@ R2Score
...
+    :header: "indexes shape", "indexes dtype", "preds shape", "preds dtype", "target shape", "target dtype"
+    :widths: 10, 10, 10, 10, 10, 10
+
+    "``long``", "(N,...)", "``float``", "(N,...)", "``long`` or ``bool``", "(N,...)"
+
+.. note::
+    All dimensions are "squeezed out" at the beginning, so
+    that, for example, a tensor of shape ``(N, M)`` is treated as ``(N * M, )``.
+
+In Information Retrieval you have a query that is compared with a variable number of documents. For each pair ``(Q_i, D_j)``,
```
Comment (SkafteNicki) [inline docs/source/references/modules.rst:287]:
```suggestion
In information retrieval you have a query that is compared with a variable number of documents. For each pair ``(Q_i, D_j)``,
```

File: docs/source/references/modules.rst [hunk 9]
```diff
@@ -269,4 +262,60 @@ R2Score
...
+(``N`` stands for the batch size) and the following types:
+
+.. csv-table::
+    :header: "indexes shape", "indexes dtype", "preds shape", "preds dtype", "target shape", "target dtype"
+    :widths: 10, 10, 10, 10, 10, 10
+
+    "``long``", "(N,...)", "``float``", "(N,...)", "``long`` or ``bool``", "(N,...)"
+
+.. note::
+    All dimensions are "squeezed out" at the beginning, so
```
Comment (SkafteNicki) [inline docs/source/references/modules.rst:284]:
```suggestion
    All dimensions are flattened at the beginning, so
```

File: docs/source/references/modules.rst (hunk 8 above)
Comment (Borda) [reply docs/source/references/modules.rst:287]: I think that this can stay if we follow shortcut like "In Information Retrieval (IR) you have a query"

**Review by ananyahjha93 — APPROVED** [review]

File: torchmetrics/functional/retrieval/average_precision.py [hunk 10]
```diff
@@ -14,40 +14,36 @@
...
         preds: estimated probabilities of each document to be relevant.
-        target: ground truth about each document being relevant or not. Requires `bool` or `int` tensor.
+        target: ground truth about each document being relevant or not.
 
     Return:
-        a single-value tensor with the average precision (AP) of the predictions `preds` wrt the labels `target`.
+        a single-value tensor with the average precision (AP) of the predictions ``preds`` wrt the labels ``target``.
 
     Example:
+        >>> import torch
```
Comment (Borda) [inline torchmetrics/functional/retrieval/average_precision.py:37]:
```suggestion
```
